server/src/test/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportBootstrapClusterActionTests.java

```diff
@@ -0,0 +1,241 @@
+            }
+
+            @Override
+            public void start() {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void stop() {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void close() {
+                throw new AssertionError("should not be called");
+            }
+        };
+        final TransportBootstrapClusterAction transportBootstrapClusterAction
+            = new TransportBootstrapClusterAction(Settings.EMPTY, mock(ActionFilters.class), transportService, discovery);
+
+        final ActionListener<AcknowledgedResponse> listener = new ActionListener<AcknowledgedResponse>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError("should not be called");
+            }
+        };
+
+        assertThat(expectThrows(IllegalStateException.class,
+            () -> transportBootstrapClusterAction.doExecute(mock(Task.class), new BootstrapClusterRequest(), listener))
+            .getMessage(), equalTo("cannot execute a Zen2 action if not using Zen2"));
+
+        threadPool.shutdown();
+    }
+
+    public void testFailsOnNonMasterEligibleNodes() {
+        final DiscoveryNode discoveryNode
+            = new DiscoveryNode("local", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT);
+
+        final MockTransport transport = new MockTransport();
+        final ThreadPool threadPool = new TestThreadPool("test", Settings.EMPTY);
+        final TransportService transportService = transport.createTransportService(Settings.EMPTY, threadPool,
+            TransportService.NOOP_TRANSPORT_INTERCEPTOR, boundTransportAddress -> discoveryNode, null, emptySet());
+        transportService.start();
+        transportService.acceptIncomingRequests();
+
+        final ClusterSettings clusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        final ClusterState state = ClusterState.builder(new ClusterName("cluster")).build();
+        final Coordinator coordinator = new Coordinator("local", Settings.EMPTY, clusterSettings, transportService,
+            ESAllocationTestCase.createAllocationService(Settings.EMPTY),
+            new MasterService("local", Settings.EMPTY, threadPool),
+            () -> new InMemoryPersistedState(0, state), r -> emptyList(),
+            new NoOpClusterApplier(), random());
+        coordinator.start();
+
+        final TransportBootstrapClusterAction transportBootstrapClusterAction
+            = new TransportBootstrapClusterAction(Settings.EMPTY, mock(ActionFilters.class), transportService, coordinator);
+
```

Comment:
There seems to be a lot of code duplication in the initialization part of the tests, which is not important for the test itself, can we refactor a method for intialization?

Reply:
I did so in fc2f125baa735f4a889a57b959337f3578588f91. I'm not sure I wholly like it, because of things like the fact that we have to change the local node and rely on the transport service not picking it up until started, but I don't have a strong opinion about this.